Update transaction builder

brennan opened the pull request:
@f2prateek @ladanazita is it possible to get a brief code review on these ecommerce updates? 

Synopsis of changes: 

Removed deprecated ecommerce methods: 
- HitBuilders.TransactionBuilder()
- HitBuilders.ItemBuilder()

All ecommerce events are now tracked with the HitBuilders.ScreenViewBuilder() method ; product objects must be attached to screen views to be sent to GA

brennan committed:
# This is a combination of 11 commits.
# The first commit's message is:
Update transaction methods

# This is the 2nd commit message:

Add orderId to ProductAction method

# This is the 3rd commit message:

Remove Transaction Build import

# This is the 4th commit message:

Updated track:completed order

# This is the 5th commit message:

Updating product event methods

# This is the 6th commit message:

Refactoring product hits

# This is the 7th commit message:

Refactoring

# This is the 8th commit message:

Fixing circleci errors

# This is the 9th commit message:

Fixing circleci style errors

# This is the 10th commit message:

Fixing tests

# This is the 11th commit message:

Fixing tests

brennan committed:
Update transaction methods

f2prateek commented on src/main/java/com/segment/analytics/android/integrations/google/analytics/GoogleAnalyticsIntegration.java:
```diff
@@ -72,7 +72,12 @@
       Pattern.compile("completed *order", CASE_INSENSITIVE);
   static final Pattern PRODUCT_EVENT_NAME_PATTERN =
-      Pattern.compile("((viewed)|(added)|(removed)) *product *.*", CASE_INSENSITIVE);
+      Pattern.compile("(((viewed)|(added)|(removed)) *product *.*)|"
```
It seems both `verb action` and `action verb` are supported - shouldn't it just be one?

ladanazita commented:
LGTM, but I think we should audit what functionality is no longer supported, such as mapping to `sku` and clearly document this and email success with the update.

f2prateek commented on src/main/java/com/segment/analytics/android/integrations/google/analytics/GoogleAnalyticsIntegration.java:
```diff
@@ -180,31 +185,41 @@
     if (COMPLETED_ORDER_PATTERN.matcher(event).matches()) {
+
+      // initialize instance of ScreenViewBuilder class
+      ScreenViewHitBuilder builder = new ScreenViewHitBuilder();
```
FWIW this does seem weird to me that `ScreenViewBuilder` is being used for this. Can you link to the docs?

brennan replied:
Here's an example of a transaction event: https://developers.google.com/analytics/devguides/collection/android/v4/enhanced-ecommerce#measuring-transactions. It seems all their enhanced ecommerce events are associated with screen views in their examples. Although I could probably attach Completed Order, Product Viewed, etc., events to event hits, seen here: https://developers.google.com/analytics/devguides/collection/android/v4/events

f2prateek replied:
I see, I do think an event would be more appropriate here. Otherwise I'm worried this would inflate screen counts for users (since previously it wouldn't create a screen hit).

brennan replied:
Agreed now that I'm understanding their API better. Will re-implement. Thanks!

brennan committed:
Updating methods

brennan committed:
Add enhanced ecommerce event actions

brennan committed:
Update tests/fix typos

brennan commented:
Ok, guys! Added test for "Product Viewed", etc., pattern and fixed typos. Han confirmed we should support both because we provide backwards ecomm spec compatibility on client/server.

ladanazita commented on src/main/java/com/segment/analytics/android/integrations/google/analytics/GoogleAnalyticsIntegration.java:
```diff
@@ -69,10 +69,16 @@
   private static final ValueMap EMPTY = new ValueMap(Collections.<String, Object>emptyMap());
   private static final String DEFAULT_CATEGORY = "All";
+  private static final String ENHANCED_ECOMMERCE = "EnhancedEcommerce";
   static final Pattern COMPLETED_ORDER_PATTERN =
```
Should we also support [order completed ](https://segment.com/docs/spec/ecommerce/v2/#order-completed) here?

ladanazita commented on src/main/java/com/segment/analytics/android/integrations/google/analytics/GoogleAnalyticsIntegration.java:
```diff
@@ -342,19 +342,40 @@ void sendProductEvent(String event, String category, Properties properties) {
-    Map<String, String> itemHit = itemHitBuilder.build();
-    tracker.send(itemHit);
-    logger.verbose("tracker.send(%s);", itemHit);
+        .setQuantity(properties.getInt(QUANTITY_KEY, 0));
```
Does GA expect type long or int?

ladanazita replied:
Discussed in person - properties.quantity is meant to be passed in as an int/number according to our spec, so this should be o.k.

ladanazita commented on src/test/java/com/segment/analytics/android/integrations/google/analytics/GoogleAnalyticsTest.java:
```diff
@@ -335,42 +338,86 @@ static AnalyticsContext contextWithCampaign(Campaign campaign) throws Exception
         .putPrice(20)
         .putValue("quantity", 10)
-        .putValue("customMetric", "32.22")
+        .putValue("customMetric", 32)
```
Did this fail with the input 32.22?

brennan committed:
Update completed order regex

f2prateek commented on src/main/java/com/segment/analytics/android/integrations/google/analytics/GoogleAnalyticsIntegration.java:
```diff
@@ -180,31 +187,42 @@
 
     if (COMPLETED_ORDER_PATTERN.matcher(event).matches()) {
+
+      // initialize instance of EventHitBuilder
```
This comment is unnecessary.

f2prateek commented on src/main/java/com/segment/analytics/android/integrations/google/analytics/GoogleAnalyticsIntegration.java:
```diff
@@ -180,31 +187,42 @@
-          hitBuilder.setTransactionId(properties.orderId())
+
+          // initialize instance of newProduct
+          com.google.android.gms.analytics.ecommerce.Product newProduct =
```
This comment is unnecessary.

f2prateek commented on src/main/java/com/segment/analytics/android/integrations/google/analytics/GoogleAnalyticsIntegration.java:
```diff
@@ -180,31 +187,42 @@
-          logger.verbose("tracker.send(%s);", hit);
+              .setQuantity(product.getInt(QUANTITY_KEY, 0));
+
+          // add product to builder object
```
This comment is unnecessary.

f2prateek commented on src/main/java/com/segment/analytics/android/integrations/google/analytics/GoogleAnalyticsIntegration.java:
```diff
@@ -180,31 +187,42 @@
+      eventHitBuilder.setProductAction(productAction)
+             .setCategory(ENHANCED_ECOMMERCE);
+
+      // below method requires first method to be of type CustomHitBuilder
```
I'm not sure what this comment means or why it's necessary.

f2prateek replied:
Do you mean first argument to be of type `CustomHitBuilder`? Is this comment necessary?

brennan committed:
Removing unnecessary comments

f2prateek commented on src/main/java/com/segment/analytics/android/integrations/google/analytics/GoogleAnalyticsIntegration.java:
```diff
@@ -180,31 +187,42 @@
+          .setTransactionShipping(properties.shipping());
+
+      eventHitBuilder.setProductAction(productAction)
+             .setCategory(ENHANCED_ECOMMERCE);
```
Shouldn't the category be what's passed in properties instead of a hardcoded string?
 https://developers.google.com/android/reference/com/google/android/gms/analytics/ecommerce/Product.html#setCategory(java.lang.String)

brennan committed:
Change EnhancedEcommerce category to fallback

brennan committed:
Fixing circleci errors
